### PR TITLE
Hw2 events.async

### PR DESCRIPTION
**Pull request by winrono**

### CODE REVIEW

File: file-handling/importer.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        //processing only non-empty lines
+        const lines = fs.readFileSync(path, 'utf8').split('\r\n').filter(Boolean);
+        const columns = lines[0].split(',');
+        let objects = [];
+
+        //skip first line as it's header
+        for (let i = 1; i < lines.length; i++) {
+            const values = lines[i].split(',');
+            const obj = {};
+            columns.forEach((column, index) => {
```
Comment (spzm) [inline file-handling/importer.js]: can be replaced with reduce

File: file-handling/dirwatcher.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+//used by alternative version
+//const statAsync = promisify(fs.stat);
+
+export class DirWatcher extends EventEmitter {
+    constructor() {
+        super();
+        this.processedFiles = [];
+    }
+
+    watch(path, delay) {
```
Comment (spzm) [inline file-handling/dirwatcher.js:18]: please add an `unwatch` method that will stop watcher

File: app.js [hunk 3]
```diff
@@ -1,7 +1,23 @@
...
\ No newline at end of file
+const product = new Product();
+
+const dirWatcher = new DirWatcher();
+dirWatcher.watch("./data", 5000);
+
+const importer = new Importer();
+
+dirWatcher.on('changed', (path) => {
+    importer.import(path).then((json) => {
```
Comment (spzm) [inline app.js:16]: please handle the situation when file is not possible to process (incorrect format) or it's a folder

File: file-handling/dirwatcher.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+        //using setTimeout approach to make sure that dirwatcher is able to handle arbitrary number of CSV files without overlapping
+        const action = async () => {
+
+            const files = await readdirAsync(path);
+
+            await Promise.all(files.map(async (filename) => {
```
Comment (spzm) [inline file-handling/dirwatcher.js]: async is not needed if there is no await inside a function

File: file-handling/importer.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            const reader = readline.createInterface(fs.createReadStream(path));
+
+            reader.on('line', (line) => {
+                const values = line.split(',');
+                //if columns is undefined then first line is processing 
+                if (!columns) {
+                    columns = values;
+                } else {
+                    const obj = {};
+                    columns.forEach((column, index) => {
```
Comment (spzm) [inline file-handling/importer.js]: reduce can be used

File: file-handling/importer.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        });
+    }
+    //async implementation without library usage
+    importAsync(path) {
+        return new Promise((resolve, reject) => {
+
+            let objects = [];
+            let columns;
+
+            const reader = readline.createInterface(fs.createReadStream(path));
```
Comment (spzm) [inline file-handling/importer.js]: please process the case when error can happen during reading process

File: file-handling/importer.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                resolve(JSON.stringify(objects));
+            });
+
+        });
+    }
+    //csvtojson doesn't provide sync ways to convert csv to json
+    //so I decided to not use yet another library and provided simple implementation instead
+    importSync(path) {
+        //processing only non-empty lines
+        const lines = fs.readFileSync(path, 'utf8').split('\r\n').filter(Boolean);
```
Comment (spzm) [inline file-handling/importer.js]: just to note to be accurate with `\r\n` - it's OS specific

**Comment by winrono** [comment]
Added fixes for mentioned issues/improvements

File: file-handling/dirwatcher.js (hunk 4 above)
Comment (winrono) [reply file-handling/dirwatcher.js]: required again after latest commit

File: file-handling/dirwatcher.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        this.processedFiles = [];
+    }
+
+    watch(path, delay) {
+        this.stopWatching = false;
+        //using setTimeout approach to make sure that dirwatcher is able to handle arbitrary number of CSV files without overlapping
+        const action = async () => {
+
+            const files = await readdirAsync(path);
+            await Promise.all(files.map(async (filename) => {
```
Comment (spzm) [inline file-handling/dirwatcher.js:24]: I don't see that in callback any promise is returned. Is `Promise.all` needed?

**Review by spzm — APPROVED** [review]

**Review by spzm — APPROVED** [review]

**Pull request merged**